Repositories

LucasEduardoGiovanini opened the pull request:
As this issue is about the repositories, i create a test which shows that it is possible to retrieve all the lists of a user within my database, as requested.
I thought of creating a class for use cases. To do operations like: create lists or retrieve them, but I didn't want get too far from the topic of that issue too much, not to skip the next steps.

LucasEduardoGiovanini committed:
creating the package structure.

LucasEduardoGiovanini committed:
creating the user repository

If run, this commit will create the package user_repository. i create a only simple def to register a user and two another functions to recovery the user list.

LucasEduardoGiovanini committed:
creating some operations in the repositories

if run, this commit will create, in todo_list_repository, some functions to make operations in database. the function "get_the_next_free_position" takes the position of the last task. this function will be used to generate a new position when the user is going to create a new task.
Also, in the user_repository, i create a task to verify if user email is registered.

LucasEduardoGiovanini committed:
making a test about insert a todo list in database.

if run, this commit will be create the the mentioned test and, in the package todo_list_repository, will create a conversion for uuid.

LucasEduardoGiovanini committed:
Creating getters and setters for task and todolist.

If run, this commit will create getters and setters for all class attributes. I create this because i need to insert the object in database, and i don't want access the attribute directly.

LucasEduardoGiovanini committed:
Creating a test to insert a task in database and fix bugs in get_next_free_position

If run, this commit will create a test to insert a task in a list. For this test, i need to extract the  list id in which the user wants to place the object and extract the next free position in this same list, to place the task. When the use cases are implemented, i wanna use a synchronized function to make ensure no one insert a task while i don't extract the next free position.

LucasEduardoGiovanini committed:
creating a test to recover all user lists.

johnnywell commented on domain/entities/task.py:
```diff
@@ -4,12 +4,37 @@
         self.completed = completed
         self.priority = priority  # informa a prioridade da task, exemplo: uma task com priridade 1 é mais importante que uma de prioridade 2
 
+    @property
```
You should not create getters and setters if you don't need it.

LucasEduardoGiovanini replied:
Can i delete it, but soon i won't need to leave my attributes private? and access them using getters and setters?

johnnywell replied:
So you do it when you need it, not before it. Got it? it, it it... rsrsrs

LucasEduardoGiovanini replied:
hahahahaha ok, thanks

johnnywell commented on domain/entities/todolist.py:
```diff
@@ -6,10 +6,26 @@
         self.name = name
         self.tasks = list()
 
+    @property
```
You should not create getters and setters you don't need them.

johnnywell commented on repositories/todo_list_repository.py:
```diff
@@ -0,0 +1,33 @@
+                                          db='dbTodoList',
+                                          cursorclass=pymysql.cursors.DictCursor)
+
+    def create_todo_list(self, email: str, id_code: str, list_name: str):
```
Repositories should return instances of your domain objects. The tests are screaming for that. I'll only make this comment about that, but please review all repositories methods.

LucasEduardoGiovanini replied:
But when I insert my object into the database, shouldn't I return what was inserted?

johnnywell replied:
Yeah, in the form of objects. If you just inserted a Todo, you should return a Todo object. Returning only the data makes the use case or whatever code used your repository have to recreate that object, and soon this piece of code will be repeatedly spread all over the place.

LucasEduardoGiovanini commented:
I think the issue is ready.
while I wait for your analysis, I'm setting up a function that generates an UPDATE query for my database. So that I can set up a dynamic query.

johnnywell commented:
I've wrote the review yesterday but I forgot to submit...

LucasEduardoGiovanini committed:
removing the not unused getters and setters.

LucasEduardoGiovanini committed:
Just renaming the attributes

If run, this commit will rename some attributes so that the table columns in database and the entities attributes to have the same name.
I make this because, when i convert the query dictionary in a object, the dictionary keys will work like object instance.

LucasEduardoGiovanini committed:
refactoring the construtor of entities.

If run, this commit will remove the creation of the ID using uuid4 directly in the constructor and receive this uuid4 as a parameter.

LucasEduardoGiovanini commented:
Now, that my application uses a database, it doesn't make any sense for my TodoList class constructor to have a list of tasks within it. So, I'm going to remove this list and add a todolist_id attribute inside the task constructor, to maintain a link between the task and a task list.

**How is it:**
```
class TodoList:
    def __init__(self, todolist_id, todo_name):
        self.todolist_id = todolist_id
        self.todo_name = todo_name
        self.tasks = list()    

class Task:
    def __init__(self, task_id, task_name, description, completed, priority):
        self.task_id = task_id
        self.task_name = task_name
        self.description = description
        self.completed = completed
        self.priority = priority
```
**how will it be**
```
class TodoList:
    def __init__(self, todolist_id, todo_name):
        self.todolist_id = todolist_id
        self.todo_name = todo_name
        s̶e̶l̶f̶.̶t̶a̶s̶k̶s̶ ̶=̶ ̶l̶i̶s̶t̶(̶)̶     <-- i remove the self.tasks

class Task:
    def __init__(self,todolist_id, task_id, task_name, description, completed, priority):
        self.todolist_id = todolist_id    <--- inserted
        self.task_id = task_id
        self.task_name = task_name
        self.description = description
        self.completed = completed
        self.priority = priority
```

LucasEduardoGiovanini committed:
modifying the returns of repository and changing the respective tests

if run, this commit will make the function create_todo_list and create_task of todo_list_repository returns a object using the dictionaries values like parameters.

johnnywell commented on repositories/todo_list_repository.py:
```diff
@@ -0,0 +1,37 @@
+        result_dictionary = cursor.fetchone()
+        return TodoList(result_dictionary['todolist_id'], result_dictionary['todo_name'])
+
+    def get_the_next_free_position(self, id_list: str):# pego a maior posição da fila
```
This doesn't work for multithread, like in a web server. Why are you doing this way instead of just save the object and let the database create the id?

johnnywell commented on repositories/user_repository.py:
```diff
@@ -0,0 +1,40 @@
+import pymysql
+
+
+class UserRepository:
```
Do you need this repository? I couldn't find it been used anywhere besides the method `get_code_from_user_list` which doesn't make any sense. I had to look this method implementation to reason what you are doing, and you surely could have make a method on the list repository to return the lists based on the email.

Tip: Take time to think about the APIs you are creating, make them like you would use, because you will.
Tip 2: Everything is an API, every variable name, every function and method name, every class.

LucasEduardoGiovanini replied:
As a user would have the list, I thought about creating a repository for him, following the same principle as the tourist spots api.
I did it thinking about using it in the future, but I will pay more attention and do things in order.

LucasEduardoGiovanini committed:
moving the uuid creation from constructor  to database.

if run, this commit will  remove the creation of id in the constructors and will start to create them in database,  automatically.
  All changes in the commit are about to remove the uuid in parameter passages. This include the Tests.

LucasEduardoGiovanini committed:
renaming a get_code_from_user_list function for refactor my test

If run, this commit will refactor the name function to make more sense with the context and refactor the tests that use this functions.

LucasEduardoGiovanini committed:
removing all dependences about the id's.

if run, this commit will remove all id parameter  from the functions create_todo_list and create_task, in package todo_list_repository. As the id is generate in database now, it is no longer needed like a parameter here.

LucasEduardoGiovanini committed:
insert one more attribute in Task and altering the return parameter.

LucasEduardoGiovanini committed:
increase a attribute in todo list

LucasEduardoGiovanini committed:
[1/2] Refactor a operation of creation in TodoList package, creating another function in repository and start the alterations about some tests.

If run, this commit will be refactor the create_task function to receives one more parameter, a todolist_id. I made this change because, in my database, my Task need a todolist_id to be created. taking this in consideration, i decided to create a function in my repository to recover a the id of todolist using the name of the list and the email of creator. Making this, i can leave the database create the ID and i can extract the id using another parameters.
But it make my system not accept a user create two or more lists with the same name or two or more tasks with the same parameters.

LucasEduardoGiovanini committed:
Applying the alterations made in the last commit

This commit is large, but is the same change for all the functions. How i create a function to recover my list id by the name and email and i increase a parameter "email" in todoList constructor, i need to pass a email now, and using the todolist datas to recover your id from the database.

LucasEduardoGiovanini commented:
I am making some changes so that the system can work with a database. So I was unable to apply all of your suggestions today. But I will implement them as soon as possible! Thanks for the suggestions and feedback, it's been really fun to implement this application hahahah

LucasEduardoGiovanini committed:
Finishing the tests step (for this moment).

LucasEduardoGiovanini committed:
Deleting the user_repository package and realocating your functions in todo_list_repository

LucasEduardoGiovanini committed:
creating a function in repository

if run, this commit will create a function to reinsert the altered task in database. Also i modified my function of create task to get the next free position automatically.

LucasEduardoGiovanini committed:
modifying the function of recover all tasks from list.

If run, this commit will make the function return a list of task objects.

LucasEduardoGiovanini committed:
refactoring a function of repository to show all lists and all tasks from user.

If run, this commit will make the function mentioned return a list of lists. For example:
all_lists_with_tasks= [ [ todolist1, [task1, task2, task3] ],[ todolist1, [task1, task2, task3]]

Also i adapt my tests for this alteration in repository.

LucasEduardoGiovanini commented:
A minha duvida é a seguinte. Está correto importar as minhas entidades dentro do meu repositório? Pois as minhas funções no repositório aparentam estar realizando muitas operações, como por exemplo: recuperar, de acordo com o email do usuário, todas as suas listas de tarefas. Faço isso utilizando os construtores das minhas entidades, que criam os objetos a serem retornados pelos métodos do repositório. Com isso eu percebo que as  entidades estão realizando menos operações que antes, parece que a minha aplicação está girando em torno do repositório, e estou achando isso um pouco estranho... 

Eu preferiria que as minhas entidades realizassem as operações, chamando os métodos do repositório, porém o meu repositório utiliza as entidades, então geraria um problema de dependência circular. 

 De qualquer forma, se semana que vem você tiver disponibilidade, podemos conversar melhor sobre. Também irei commitar meu banco de dados junto, caso queira dar uma olhada.
Obrigado pela ajuda ;)

LucasEduardoGiovanini committed:
inserting the database file and refactor a test

the test about create a list was done in a way that i don't  access the entities directly.

LucasEduardoGiovanini committed:
removing a task list from todolist.

if run, this commit wll remove the attribute "tasks" from todolist package. I made this because now the objects are stored in database. Into the tbTask, i pass the Todolist Id like a Foreign key, to relate the task to the todolist. Using this method, the attribute "tasks" it is no longer necessary to relate tasks to a list.

LucasEduardoGiovanini committed:
refactoring the test about: complete a task that is incomplete

if run, this commit will refactor the mentioned test to insert the changes into the database. I enjoy this moment to replace the character "AND" for "," in my repository.

LucasEduardoGiovanini commented:
I am now redoing some tests to test my repository operations. Maybe ending this, I can think of a better way to solve my problem.

LucasEduardoGiovanini committed:
refactoring the test about: create a task in todo list

If run, this commit will refactor the mentioned test to verify if a task exist in list using a method existent in todolist package. This method use rich comparison to check if the objects is equal.

LucasEduardoGiovanini committed:
removing the old and outdated tests.

johnnywell commented:
> A minha duvida é a seguinte. Está correto importar as minhas entidades dentro do meu repositório? Pois as minhas funções no repositório aparentam estar realizando muitas operações, como por exemplo: recuperar, de acordo com o email do usuário, todas as suas listas de tarefas.

Sim está correto, as dependências devem apontar sempre em direção ao centro da sua arquitetura. A idea central disso é que seus objetos de domínio, que são puros, não dependam da infraestrutura que os rodeiam. Independente da tecnologia que você utilizar nas camadas exteriores isso não deve impactar seu domínio.

![CleanArchitecture](https://user-images.githubusercontent.com/1902036/81060951-d907b680-8ea9-11ea-8be8-245a1d122f24.jpg)

> Faço isso utilizando os construtores das minhas entidades, que criam os objetos a serem retornados pelos métodos do repositório. Com isso eu percebo que as entidades estão realizando menos operações que antes, parece que a minha aplicação está girando em torno do repositório, e estou achando isso um pouco estranho...

Usar essa arquitetura numa aplicação tão simples é *overkill* mesmo, só o fazemos para estudar. Como existem poucas regras de negócio realmente pode ser difícil enxergar claramente onde terminam as responsabilidades das entidades e onde começam as da infraestrutura. Mas resumidamente, as Entidades não têm que lidar com nada que não sejam regras de negócio. Seus repositórios  existem para abstrair a complexidade de acesso ao banco de dados, se é isso que eles estão fazendo, então estão corretos.

> Eu preferiria que as minhas entidades realizassem as operações, chamando os métodos do repositório, porém o meu repositório utiliza as entidades, então geraria um problema de dependência circular.

Lembre-se da direção da dependência. É muito raro uma entidade precisar utilizar um repositório, mas eu consigo imaginar que existam casos em que isso é necessário. Não numa aplicação tão simples como esta. No caso de uma entidade utilizar um repositório essa referência deveria ser indireta através de uma interface e injetada durante a construção da sua entidade. Assim você não cria uma dependência nem importações circulares.

Leia o [artigo do Uncle Bob sobre Clean Architecture](https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html). Esse material não é profundo no tema, para isso você precisaria ler o livro completo, mas já dá uma boa ideia de como as coisas devem ser organizadas. Alías, minha recomendação é: Estude. sempre que vir um termo ou conceito que ainda não entende completamente mergulhe nisso e só volte a escrever código quando tiver certeza do que estiver fazendo. Quando comparado com pesquisa, escrever código é a menor parte da nossa atividade. Não estou dizendo que você não deva dedicar algum tempo para explorar ideias ou testar coisas, mas isso é interessante quando você sabe onde quer chegar. Explorar sem objetivo é perda de tempo.
